### DISCUSSION TITLE
Error in step tidy: "Error Node (/Unsqueeze) has input size 2 not in range [min=1, max=1]"

### DISCUSSION THREAD

**Discussion by MFJI** [Xilinx/finn#802] [Q&A] [answered]
Hello everyone,

I trained a model through brevitas and exported the onnx file to FINN. However, when I generated the onnx file with FINN, I met the following error in the step tidy:

"
Traceback (most recent call last):
  File "/workspace/finn/src/finn/builder/build_dataflow.py", line 168, in build_dataflow_cfg
    model = transform_step(model, cfg)
  File "<ipython-input-12-0481ba0590cb>", line 93, in step_resnet50_tidy
    model = model.transform(FoldConstants())
  File "/workspace/qonnx/src/qonnx/core/modelwrapper.py", line 140, in transform
    (transformed_model, model_was_changed) = transformation.apply(transformed_model)
  File "/workspace/qonnx/src/qonnx/transformation/fold_constants.py", line 96, in apply
    oxe.execute_node(n, execution_context, graph)
  File "/workspace/qonnx/src/qonnx/core/onnx_exec.py", line 73, in execute_node
    sess = rt.InferenceSession(node_model.SerializeToString())
  File "/opt/conda/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 335, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/opt/conda/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 372, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. In Node, ("/Unsqueeze", Unsqueeze, "", -1) : ("/Gather_output_0": tensor(int64),"onnx::Unsqueeze_787": tensor(int64),) -> ("/Unsqueeze_output_0": tensor(int64),) , Error Node (/Unsqueeze) has input size 2 not in range [min=1, max=1].

> /opt/conda/lib/python3.8/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py(372)_create_inference_session()
    370             sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
    371         else:
--> 372             sess = C.InferenceSession(session_options, self._model_bytes, False, self._read_config_from_model)
    373 
    374         if disabled_optimizers is None:
"
The version I use is
brevitas==0.8.0
torch==2.0.0
torchvision==0.15.1
onnx==1.13.1
onnxoptimizer==0.3.10
FINN==v0.9

The code I use to export the onnx file from brevitas is
import brevitas.onnx as bo
bo.export_finn_onnx(model.module.cpu(),torch.randn(1,3,224,224),'./xxx.onnx')

Does anyone know what this problem may relate to?

Thank you in advance.

**Comment by MFJI** [comment, marked as answer]
It seems like a version problem. If I use brevitas==0.7.1, torch==1.8.2, torchvision==0.9.2, onnx==1.11.0,onnxoptimizer==0.2.0 and use code bo.export_finn_onnx(model.module.cpu(), (1,3,224,224),'./xxx.onnx') to export and it works. I am considering it as a torch version problem. I will try whether a higher version of brevitas and onnx works or not.

**Reply by fpjentzsch** [reply]
Hi, such an ONNXRuntimeError can indeed point to a version incompatibility. More specifically, Brevitas exports the model with an older opset version, but (after modification by FINN) it contains newer operator variants that are incompatible with the specified opset.

An alternative workaround can be to force a newer opset version during export via the `opset_version` parameter of the `export_qonnx` function.
